### PR TITLE
Capabilities control in yml

### LINKED ISSUE #505: All capabilities off by default
"TypeScript Approach" to DevPool Offering

TypeScript can be added to a project without enabling all the features. Makes it easier to ease into using it as a beginner. 

We should do the same with UbiquiBot and start with features/incentives disabled, and then let people discover and enable the cool features as they need them.

This is because we currently have a pretty heavy handed approach with how to run organizations which can make it less attractive vs adding the bot for a single "killer feature".

---

I wish that we could ask the user if they want to initialize a repository with the recommended default settings, or start from scratch. Otherwise we should create a "recommended default configuration" `.yml` that we have in our docs for partners to copy and add in.

---

Update: here's [a checklist](https://github.com/ubiquity/ubiquibot/issues/505#issuecomment-1636708093) for the bounty.

### PR DESCRIPTION
**Pull request by devpanther**
Resolves #505 

Sample yml config here: https://github.com/one-two-test/one/blob/main/.github/ubiquibot-config.yml
Disabling Test: https://github.com/one-two-test/one/issues/1#issuecomment-1655711155

### CODE REVIEW

File: src/configs/price.ts
Comment (pavlovcik) [inline src/configs/price.ts:7]: Everything should be be off by default.

File: src/handlers/comment/action.ts
Comment (pavlovcik) [inline src/handlers/comment/action.ts:41]: I think it makes sense to name this variable "feature"

File: src/handlers/comment/handlers/index.ts
Comment (pavlovcik) [inline src/handlers/comment/handlers/index.ts:76]: We should add a log for this instead of just a silent return

**Review by pavlovcik — CHANGES_REQUESTED** [review]

File: src/handlers/comment/action.ts
Comment (devpanther) [reply src/handlers/comment/action.ts:41]: Yea, it does.. :)

File: src/configs/price.ts
Comment (devpanther) [reply src/configs/price.ts]: > Everything should be be off by default.

Okay then, I disabled em all

**Review by pavlovcik — APPROVED** [review]

**Comment by devpanther** [comment]
Huge conflict here as well :(

**Comment by devpanther** [comment]
Took me an hour to resolve this conflict, QA should be started @Draeieg

**Comment by 0xcodercrane** [comment]
> Took me an hour to resolve this conflict, QA should be started @Draeieg

Thank you for resolving conflicts. If @Draeieg has no bandwidth, @wannacfuture would you please confirm the QA?

**Comment by 0xcodercrane** [comment]
https://github.com/ubiquity/ubiquibot/pull/561

The above PR requires renaming assign -> start, unassign -> stop.

**Comment by devpanther** [comment]
> https://github.com/ubiquity/ubiquibot/pull/561
> 
> The above PR requires renaming assign -> start, unassign -> stop.

Isn't it supposed to change automatically on merge or show up if there's a conflict?

**Comment by devpanther** [comment]
@0xcodercrane resolved

**Comment by devpanther** [comment]
@0xcodercrane resolved, tested too

I see /assign and /unassign was renamed, fixed that too

**Comment by devpanther** [comment]
@0xcodercrane This json implementation you made just introduced an issue

https://stackoverflow.com/a/58941798/11664341

typescript now copies the `/src` folder into the `/lib` instead of just copying the files because you required a file outside the src folder (the json config)

**Comment by devpanther** [comment]
Screenshot 2023-08-02 at 10 46 01 AM

Here is how it looks like now

**Comment by 0xcodercrane** [comment]
Which error did you have @devpanther with the current implementation?

**Comment by 0xcodercrane** [comment]
> typescript now copies the /src folder into the /lib instead of just copying the files because you required a file outside the src > folder (the json config)

I don't catch any hints about the issue in your words.

**Comment by devpanther** [comment]
Read the stackoverflow and try running it locally yourself by deleting lib and running `tsc` again

**Comment by devpanther** [comment]
@0xcodercrane does the QA code update on code pushes, cause I just did a series of test and it seems to not respond to the changes I made

**Comment by 0xcodercrane** [comment]
I think I understand what you're saying. That error happens when you try to run locally, but not in staging/production. because it works a little bit differently. Its a serverless function on staging/production while a local bot is a nodejs app.

ok let me refactor to make it in both mode.

File: src/handlers/comment/action.ts
Comment (0xcodercrane) [inline src/handlers/comment/action.ts:41]: Not a good way to use hardcoded `/help` which has been defined in the command list.

**Comment by devpanther** [comment]
Can this PR be checked?, the QA seems unresponsive too.. Looks like there's something wrong, it is not running this branch's code

File: src/handlers/comment/action.ts
Comment (0xcodercrane) [inline src/handlers/comment/action.ts:44]: Why `continue`?  How will it work?

**Review by 0xcodercrane — CHANGES_REQUESTED** [review]

File: src/handlers/comment/action.ts
Comment (0xcodercrane) [reply src/handlers/comment/action.ts:41]: you might need to figure out to resolve conversations.

File: src/handlers/comment/action.ts
Comment (0xcodercrane) [reply src/handlers/comment/action.ts:44]: right. I was not looking at the `for` loop.

File: src/handlers/comment/action.ts
Comment (devpanther) [reply src/handlers/comment/action.ts]: > you might need to figure out to resolve conversations.

Figured it out

**Comment by 0xcodercrane** [comment]
> I think this would work, just wanted @0xcodercrane to understand the issue and approve its use

Lets do that! @devpanther @wannacfuture

File: src/handlers/comment/action.ts
Comment (0xcodercrane) [reply src/handlers/comment/action.ts]: You need to post the proof before marking this conversation resolved.

Comment (devpanther) [reply src/handlers/comment/action.ts]: I pushed a commit and then resolved em

**Comment by devpanther** [comment]
> > I think this would work, just wanted @0xcodercrane to understand the issue and approve its use
> 
> Lets do that! @devpanther @wannacfuture

Added the fix

**Comment by 0xcodercrane** [comment]
Waiting for QA approval.

**Comment by devpanther** [comment]
> Waiting for QA approval.

I think the QA needs to be set up again, looks broken

**Review by Draeieg — APPROVED** [review]

File: ubiquibot-config-default.yml
Comment (0xcodercrane) [inline ubiquibot-config-default.yml:1]: Why default yml still here?

**Review by 0xcodercrane — CHANGES_REQUESTED** [review]

File: ubiquibot-config-default.yml
Comment (devpanther) [reply ubiquibot-config-default.yml]: > Why default yml still here?

Hmm, weird.. i have removed it now

File: ubiquibot-config-default.json
Comment (pavlovcik) [inline ubiquibot-config-default.json:84]:
```suggestion
```

File: ubiquibot-config-default.json
Comment (pavlovcik) [inline ubiquibot-config-default.json:87]:
```suggestion
```

This should be the only hidden property so that we don't remind partners to opt-out, basically.

File: ubiquibot-config-default.json
Comment (pavlovcik) [inline ubiquibot-config-default.json:86]:
```suggestion
  "auto-pay-mode": false,
```

File: ubiquibot-config-default.json
Comment (pavlovcik) [inline ubiquibot-config-default.json:84]:
```suggestion
  "default-labels": [],
```

File: ubiquibot-config-default.json
Comment (pavlovcik) [inline ubiquibot-config-default.json:1]: I can't directly suggest on the earlier lines but I think we should do the following defaults:

```yml
price-multiplier: 1
issue-creator-multiplier: 2 # we need to convert this formula to work with 2, not 2000
time-labels: ... # these should be dynamically generated 

```

**Review by pavlovcik — CHANGES_REQUESTED** [review]

File: ubiquibot-config-default.json
Comment (devpanther) [reply ubiquibot-config-default.json:1]: > I can't directly suggest on the earlier lines but I think we should do the following defaults:
> 
> ```yaml
> price-multiplier: 1
> issue-creator-multiplier: 2 # we need to convert this formula to work with 2, not 2000
> time-labels: ... # these should be dynamically generated 
> ```

I guess it'd be taken care of on the new issue. nothing more on this PR and QA is passed

File: package.json
Comment (pavlovcik) [inline package.json:24]: Is this correct

**Review by pavlovcik — APPROVED** [review]

**Review by wannacfuture — APPROVED** [review]

File: package.json
Comment (devpanther) [reply package.json:24]: Yes

**Review by 0xcodercrane — APPROVED** [review]

File: ubiquibot-config-default.json
Comment (pavlovcik) [inline ubiquibot-config-default.json:88]: I just realized that perhaps this should be a value that represents no limit. Technically because this isn't a JavaScript file, we can't use `Infinity` but some substitute would make sense.

Comment (rndquu) [reply ubiquibot-config-default.json:88]: What [properties](https://github.com/devpanther/bounty-bot/blob/350f404e8d98f080176079122c3e2246fdf72df8/ubiquibot-config-default.json) exactly should allow the `Infinity` word? I don't see any where `Infinity` is applicable.

Comment (pavlovcik) [reply ubiquibot-config-default.json:88]: `max-concurrent-assigns: Infinity`

Comment (rndquu) [reply ubiquibot-config-default.json:88]: I don't really like it because under the hood in the js code we would need to transform this `Infinity` to either `undefined` either `Number.MAX_SAFE_INTEGER` which adds unnecessary code complexity.

Comment (pavlovcik) [reply ubiquibot-config-default.json:88]: We should do something or set it as `undefined` to represent no limits though.

Comment (rndquu) [reply ubiquibot-config-default.json:88]: If we set `max-concurrent-assigns` to `undefined` then at some point in the code we would need to write smth like:
```
const bountyHunterMax = getBountyHunterMaxFromConfig() ?? Number.MAX_SAFE_INTEGER;
```

Perhaps [here](https://github.com/ubiquity/ubiquibot/blob/development/ubiquibot-config-default.json#L89) we can simply set the default value for the `max-concurrent-assigns` parameter to be `Number.MAX_SAFE_INTEGER (9007199254740991)`

This way there are (almost) no limits for max number of assigned issues unless a partner project sets its own value for the `max-concurrent-assigns` parameter

Comment (pavlovcik) [reply ubiquibot-config-default.json:88]: Sure thing! We should also include a comment that explains why we used that integer.

**Pull request merged**
